designs/inline-io.md

```diff
@@ -0,0 +1,207 @@
+# Inline Operation Inputs / Outputs
+
+This document describes a way to write input and output shapes as an inline
+part of an operation’s definition.
+
+## Motivation
+
+Operation input and output shapes are always structures, almost always have
+boilerplate names, and critically are almost never re-used. In some cases, they
+may not even have types generated for them. Because of those properties, the
+need to fully define them separately from an operation can feel like needless
+boilerplate. Additionally, separating them makes reading an operation at a high
+level more difficult since you have to jump around to get the information you
+need.
+
+## Proposal
+
+Operations will allow inlining input and output definitions, indicated by a
+walrus operator (`:=`). Inlined structures will deviate from normal
+structure definitions in two respects. Firstly, the structure keyword will be
+omitted. Additionally, providing a name for the inlined shape will be optional.
+
+### Walrus Operator
+
+Rather than using the same standalone colon (`:`) that is used in other
+cases, a walrus operator will be used to indicate an inline definition. The
+reason for this is to visually distinguish it at the outset, as well as to make
+tooling a bit cleaner to write.
+
+This usage of the operator is analogous to how some programming languages use
+it. For instance, Python uses it to assign a name to the result of an
+expression in places where initializing a variable was previously forbidden.
+In Go, it’s used to initialize and assign a variable in one step.
+
+### Omitting the structure Keyword
+
+When used at the top level of a Smithy IDL file, the type keywords are
+necessary to indicate what type of shape you’re making. Since operation
+inputs and outputs may only be structures, this isn’t necessary.
+
+### Default Names
+
+Inlined structures will have names generated for them if they aren’t
+provided. For inputs, the default name will be the name of the operation with
+an Input suffix. For outputs, the default name will be the name of the
+operation with an Output suffix.
+
+The reason for providing a default name is that there’s rarely a better name
+than what is trivially generated. Therefore, requiring users to write it out
+is effectively pointless boilerplate. In AWS models, for instance, over 98% of
+operation input/output structures are named by suffixing the operation name.
+
+### Examples
+
+```
+operation GetUser {
+    // This uses the derived name: GetUserInput
+    input := {
```

Comment:
Given we're adding `@input` and `@output` traits to Smithy, this new syntax should automatically apply the `@input` and `@output` traits to inline defined shapes. Given that, I'm kind of liking this syntax too:

```
operation GetUser {
    @input: {
        userId: String
    }
    @output: {
        userId: String
    }
    errors: []
}
```

If we wanted to extend this inline syntax out to other places, we could just say that `@` as a prefix is the convention for defining inline properites on shapes. It's just as generic as `:=` IMO.

Reply:
`@` has a cognitive conflict with trait application though. While it's true that we may automatically apply those traits, it isn't what is most important to call out visually. Or, it's not so important that it makes it worth imposing the extra cognitive load.